dcp-0011/dcp-0011.mediawiki

```diff
@@ -11,5 +11,1259 @@ Requires: DCP0001, DCP0005, DCP0006, DCP0010
+0x007fffff may be applied to extract this value.
+|-
+|23||Sign bit.  The encoded value MUST be interpreted as negative when this bit
+is set.
+|-
+|24-31||Exponent.  The base 256 exponent plus 3.
+|}
+
+====Implementation Warning====
+
+The provided formulas for encoding specify the required calculations precisely,
+however, most programming languages will not produce the correct results for all
+values when naively implementing the formulas directly due to rounding errors
+and precision constraints in libraries that implement logarithms
+<sup>[ [[#typical-logarithm-precision-limitations|Details]] ]</sup>.
+
+Therefore, it is HIGHLY RECOMMENDED to use the algorithms outlined in the
+remainder of this section to perform the encoding and decoding versus attempting
+to implement the formulas directly.  The provided algorithms implement the
+formulas with a combination of integer math, modular arithmetic, bit shifts, and
+bitmasks, all of which produce exact results and are highly portable.
+
+====Precision Considerations====
+
+The following algorithms rely on arbitrary precision integers in order to
+support decoding the full range of possible values that can be encoded.
+
+However, in practice, all target difficulties in Decred are enforced to be
+unsigned 256-bit integers.
+
+Due to that, implementations MAY alternatively use unsigned 256-bit integers so
+long as they properly detect and reject encoded values that are negative or are
+greater than 2<sup>256</sup> - 1, the maximum value representable by an unsigned
+256-bit integer, when decoding.
+
+====Decoding Algorithm====
+
+The following algorithm will decode a target difficulty <code>T</code> from
+difficulty bits <code>c</code> using intermediate variables <code>m</code> for
+the mantissa and <code>x</code> for the exponent:
+
+# Set <code>m = c (mod 2^23)</code> so that it is the least significant 23 bits of <code>c</code>
+#: Implementations MAY optimize this by using a bitmask: <code>m = c & 0x007fffff</code>
+# Set <code>x = floor(c / 2^24)</code> so that it is the most significant 8 bits of <code>c</code>
+#: Implementations MAY optimize this by using a right bit shift: <code>x = c >> 24</code>
+# If <code>x ≤ 3</code> then set <code>T = m / 256^(3 - x)</code> otherwise set <code>T = m * 256^(x - 3)</code>
+#: Implementations MAY optimize the <code>x ≤ 3</code> case by using a right bit shift: <code>T = m >> (8 * (3-x))</code> and the <code>x > 3</code>case by using a left bit shift: <code>T = m << (8 * (x-3))</code>
+#: NOTE: Ensure <code>T</code> is capable of handling big integers
+# If bit 23 of <code>c</code> is set then return <code>-T</code> otherwise return <code>T</code>
+
+====Encoding Algorithm====
+
+The following algorithm will encode a target difficulty <code>T</code> to
+difficulty bits <code>c</code> using intermediate variables <code>m</code> for
+the mantissa and <code>x</code> for the exponent:
+
+# OPTIONAL: Assert <code>T ≥ 0</code>
```

Comment:
Fair point.  If asserting non-negative, it's not a bad idea to assert the upper end of the range too.

Updated to:

> OPTIONAL: Assert <code>T ≥ 0</code> and <code>T ≤ 2^256 - 1</code>
>  Implementations MAY optionally assert the input value is non-negative and is less than or equal to the max value representable by an unsigned 256-bit integer since target difficulties outside of that range are invalid in Decred